File: rust/do/src/nightup/go.rs
use regex::Regex;
use std::fs;
use std::path::Path;
use std::process::{Command, ExitCode};

use crate::utils;

pub fn run(dist_dir: &str, download_dir: &str) -> ExitCode {
    // index.jsonを取得する
    let output = match Command::new("curl")
        .args(["-sSL", "-A", "Mozilla/5.0", "https://go.dev/dl/?mode=json"])
        .output()
    {
        Ok(out) => out,
        Err(_) => {
            eprintln!("failed to 'curl'");
            return ExitCode::FAILURE;
        }
    };
    println!("Download (json) is done");

    let contents = match String::from_utf8(output.stdout) {
        Ok(contents) => contents,
        Err(_) => {
            eprintln!("failed to 'String::from_utf8'");
            return ExitCode::FAILURE;
        }
    };

    // ダウンロードURLを取得する
    let re_url = Regex::new(r#""filename":\s*"(go[0-9.]+\.windows-amd64\.zip)""#)
        .expect("failed to compile go.re_url (regex)");

    // 抽出できた URL を格納する変数
    let mut filename = String::new();

    if let Some(caps) = re_url.captures(&contents)
        && let Some(mat) = caps.get(1)
    {
        filename = mat.as_str().to_string();
    }

    if filename.is_empty() {
        eprintln!("failed to find ZIP URL for go-windows-amd64");
        return ExitCode::FAILURE;
    }

    let download_url = format!("https://go.dev/dl/{}", filename);

    println!("Download URL: {}", download_url);

    let work_dir_name = "go-latest-upgrade-working";
    let work_dir_path = Path::new(download_dir).join(work_dir_name);

    if work_dir_path.exists() {
        if let Err(e) = fs::remove_dir_all(&work_dir_path) {
            eprintln!("failed to remove {}: {}", work_dir_path.display(), e);
            return ExitCode::FAILURE;
        }
        println!(r#"Removed: "{}""#, work_dir_path.display());
    }

    if let Err(e) = fs::create_dir_all(&work_dir_path) {
        eprintln!("failed to create {}: {}", work_dir_path.display(), e);
        return ExitCode::FAILURE;
    }
    println!(r#"Created: "{}""#, work_dir_path.display());

    // これ以降、どこで `return` しても `work_dir_path` は自動削除される
    let mut cleanup = utils::CleanupDir {
        path: work_dir_path.clone(),
        active: true,
    };

    let local_zip = "go-latest.zip";

    let output = match Command::new("curl")
        .current_dir(&work_dir_path)
        .args(["-fsSL", "-A", "Mozilla/5.0", &download_url, "-o", local_zip])
        .output()
    {
        Ok(out) => out,
        Err(_) => {
            eprintln!("failed to 'curl ZIP file'");
            return ExitCode::FAILURE;
        }
    };

    if !output.status.success() {
        eprintln!("curl failed with status: {:?}", output.status);
        return ExitCode::FAILURE;
    }
    println!("Download (ZIP) is done");

    let output = match Command::new("tar")
        .current_dir(&work_dir_path)
        .args(["-xf", local_zip, "--strip-components=1"])
        .output()
    {
        Ok(out) => out,
        Err(_) => {
            eprintln!("failed to run OS native 'tar'. Make sure Windows 10/11 is updated.");
            return ExitCode::FAILURE;
        }
    };

    if !output.status.success() {
        eprintln!("tar failed with status: {:?}", output.status);
        return ExitCode::FAILURE;
    }
    println!("Extraction is done");

    // 後始末と配置
    let zip_path = work_dir_path.join(local_zip);
    if zip_path.exists() {
        if let Err(e) = fs::remove_file(&zip_path) {
            eprintln!("failed to remove {}: {}", zip_path.display(), e);
            return ExitCode::FAILURE;
        }
        println!(r#"Removed: "{}""#, zip_path.display());
    }

    // 配置（アップデートの適用）
    let target_path = Path::new(dist_dir);
    // 既存の古いインストールディレクトリがあれば削除
    if target_path.exists() {
        if let Err(e) = fs::remove_dir_all(target_path) {
            eprintln!("failed to remove old dist directory: {}", e);
            return ExitCode::FAILURE;
        }
        println!(r#"Removed: "{}""#, target_path.display());
    }

    // 中身がファイル群だけになったワークスペースそのものを
    // そのまま dist_dir のパスへ移動リネームする
    if let Err(e) = fs::rename(&work_dir_path, target_path) {
        eprintln!("failed to move extracted directory to dist: {}", e);
        return ExitCode::FAILURE;
    }

    cleanup.active = false; // 移動に成功したため、Drop時の削除を免除する

    println!(
        r#"Moved: "{}" to "{}""#,
        work_dir_path.display(),
        target_path.display()
    );

    println!(r#"Updated: "{}""#, dist_dir);

    ExitCode::SUCCESS
}

File: nim/do/src/nightup/go.nim
import std/[os, osproc, strformat]
import regex

import ../[utils]

proc run*(distDir: string, downloadDir: string) =
  # 1. 最新バージョンのJSONを取得
  let url = "https://go.dev/dl/?mode=json"
  let (output, curlRes) = execCmdEx(fmt"""curl -sSL -A "Mozilla/5.0" "{url}"""")
  if curlRes != 0:
    stderrMsgAndExit "failed to download json"
  
  echo "Download (json) is done"

  # 2. 正規表現で Windows 用の ZIP ファイル名を抽出
  let pattern = re2("""filename":\s*"(go[0-9.]+\.windows-amd64\.zip)""")
  var m = RegexMatch2()
  let match = find(output, pattern, m)

  var filename = ""
  if match:
    filename = output[m.group(0)]

  if filename == "":
    stderrMsgAndExit "failed to find ZIP URL for go-windows-amd64"
  
  let downloadUrl = "https://go.dev/dl/" & filename
  echo "Download URL: ", downloadUrl

  # 3. 作業用ディレクトリの作成
  let workDirName = "go-latest-upgrade-working"
  let workDirPath = downloadDir / workDirName

  try:
    rmDirIfExist(workDirPath)
    echo fmt"""Removed: "{workDirPath}""""
  except CatchableError:
    stderrMsgAndExit fmt"failed to remove existing work dir: {workDirPath}"

  try:
    createDir(workDirPath)
    echo fmt"""Created: "{workDirPath}""""
  except CatchableError:
    stderrMsgAndExit fmt"failed to create work dir: {workDirPath}"

  # 4. ZIPファイルのダウンロード
  let localZip = "go-latest.zip"
  let localZipPath = workDirPath / localZip

  let zipProcess = startProcess(
    "curl",
    args = ["-fsSL", "-A", "Mozilla/5.0", downloadUrl, "-o", localZip],
    workingDir = workDirPath,
    options = {poUsePath, poParentStreams}
  )
  let zipExit = zipProcess.waitForExit()
  zipProcess.close()

  if zipExit != 0:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit "failed to download ZIP"
  
  echo "Download (ZIP) is done"

  # 5. 外部コマンド tar の実行
  let tarProcess = startProcess(
    "tar",
    args = ["-xf", localZip, "--strip-components=1"],
    workingDir = workDirPath,
    options = {poUsePath, poParentStreams}
  )
  let tarExit = tarProcess.waitForExit()
  tarProcess.close()

  if tarExit != 0:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit "failed to extract ZIP"
  
  echo "Extraction is done"

  # 6. 不要になったZIPの削除
  try:
    rmIfExist(localZipPath)
    echo fmt"""Removed: "{localZipPath}""""
  except CatchableError:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit "failed to remove local ZIP file"

  # 7. 配置（アップデートの適用）
  try:
    rmDirIfExist(distDir)
    echo fmt"""Removed: "{distDir}""""
  except CatchableError:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit fmt"failed to remove distDir: {distDir}"

  # ワークスペースを作業パスから distDir へ移動
  try:
    moveDir(workDirPath, distDir)
    echo fmt"""Moved: "{workDirPath}" to "{distDir}""""
    echo fmt"""Updated: "{distDir}""""
  except CatchableError:
    rmDirIfExist(workDirPath)
    stderrMsgAndExit fmt"failed to move directory to {distDir}"
